### PR TITLE
feat(connector): add dispute webhooks for Stripe

### LINKED ISSUE #967: [FEATURE]  unix timestamp to primitive date time
### Feature Description

add a deserialization logic to convert unix timestamp to primitive date time so that there will no need to timestamp conversion in connectors

### Possible Implementation

need to implement a new trait which has implementations of converting unix timestamp to required date time format 

### Have you spent some time to check if this feature request has been raised before?

- [X] I checked and didn't find similar issue

### Have you read the Contributing Guidelines?

- [X] I have read the [Contributing Guidelines](https://github.com/juspay/hyperswitch/blob/main/docs/CONTRIBUTING.md)

### Are you willing to submit a PR?

Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by ShankarSinghC**
## Type of Change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates

## Description
Implemented Dispute Webhooks for Stripe

### Additional Changes

- [x] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## Motivation and Context

Fixes #967.

## How did you test it?
I tested this change manually

Out going webhook to merchant
Screenshot 2023-04-19 at 1 15 10 PM

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

### CODE REVIEW

File: crates/router/src/connector/stripe.rs [hunk 1]
```diff
@@ -1026,23 +1027,70 @@ impl api::IncomingWebhook for Stripe {
             .parse_struct("StripeWebhookObjectId")
             .change_context(errors::ConnectorError::WebhookReferenceIdNotFound)?;
 
-        Ok(api_models::webhooks::ObjectReferenceId::PaymentId(
-            api_models::payments::PaymentIdType::ConnectorTransactionId(details.data.object.id),
-        ))
+        Ok(match details.data.object.object.as_str() {
```
Comment (jagan-jaya) [inline crates/router/src/connector/stripe.rs:1030]: details.data.object.object doesn't tell what data you are accessing, give better naming here for each field

File: crates/router/src/connector/stripe.rs [hunk 2]
```diff
@@ -1026,23 +1027,70 @@ impl api::IncomingWebhook for Stripe {
...
+            "payment_intent" => api_models::webhooks::ObjectReferenceId::PaymentId(
```
Comment (jagan-jaya) [inline crates/router/src/connector/stripe.rs:1031]: change "payment_intent" and "dispute" into enum and let the parsing handle enum

File: crates/router/src/connector/stripe.rs [hunk 3]
```diff
@@ -1026,23 +1027,70 @@ impl api::IncomingWebhook for Stripe {
...
+        let webhook_obj: stripe::StripeWebhookObjectEventType = request
             .body
             .parse_struct("StripeWebhookObjectEventType")
             .change_context(errors::ConnectorError::WebhookEventTypeNotFound)?;
-
-        Ok(match details.event_type.as_str() {
-            "payment_intent.payment_failed" => api::IncomingWebhookEvent::PaymentIntentFailure,
-            "payment_intent.succeeded" => api::IncomingWebhookEvent::PaymentIntentSuccess,
+        let dispute_payload = match details.data.object.object.as_str() {
+            "dispute" => {
```
Comment (jagan-jaya) [inline crates/router/src/connector/stripe.rs:1060]: Change this to enum , let the parsing throw error for unknown types

File: crates/router/src/connector/stripe/transformers.rs [hunk 4]
```diff
@@ -1333,7 +1340,86 @@ pub struct StripeWebhookObjectResource {
 #[derive(Debug, Deserialize)]
 pub struct StripeWebhookObjectEventType {
     #[serde(rename = "type")]
-    pub event_type: String,
+    pub event_type: WebhookEvenType,
+    pub data: serde_json::Value,
```
Comment (jagan-jaya) [inline crates/router/src/connector/stripe/transformers.rs:1344]:
```suggestion
    pub data: Option<DataObject>,
```

give better naming

File: crates/router/src/connector/stripe/transformers.rs [hunk 5]
```diff
@@ -1333,7 +1340,86 @@ pub struct StripeWebhookObjectResource {
...
+}
+
+#[derive(Debug, Deserialize, PartialEq)]
+pub struct DataObject {
+    pub object: ObjectData,
+}
```
Comment (jagan-jaya) [inline crates/router/src/connector/stripe/transformers.rs:1350]: Give better naming

File: crates/router/src/connector/stripe/transformers.rs [hunk 6]
```diff
@@ -1333,7 +1340,86 @@ pub struct StripeWebhookObjectResource {
 #[derive(Debug, Deserialize)]
 pub struct StripeWebhookObjectEventType {
     #[serde(rename = "type")]
-    pub event_type: String,
+    pub event_type: WebhookEvenType,
+    pub data: serde_json::Value,
+}
+
+#[derive(Debug, Deserialize, PartialEq)]
+pub struct DataObject {
```
Comment (jagan-jaya) [inline crates/router/src/connector/stripe/transformers.rs:1348]: Give better naming

File: crates/router/src/connector/stripe/transformers.rs [hunk 7]
```diff
@@ -1333,7 +1340,86 @@ pub struct StripeWebhookObjectResource {
...
+    pub object: ObjectData,
+}
+
+#[derive(Debug, Deserialize, PartialEq)]
+pub struct ObjectData {
```
Comment (jagan-jaya) [inline crates/router/src/connector/stripe/transformers.rs:1353]: change the name

**Review by jagan-jaya — CHANGES_REQUESTED** [review]

File: crates/router/src/connector/stripe.rs [hunk 8]
```diff
@@ -1026,23 +1027,70 @@ impl api::IncomingWebhook for Stripe {
...
+            "dispute" => api_models::webhooks::ObjectReferenceId::PaymentId(
+                api_models::payments::PaymentIdType::ConnectorTransactionId(
+                    details
+                        .data
+                        .object
+                        .payment_intent
+                        .ok_or(errors::ConnectorError::WebhookReferenceIdNotFound)?,
+                ),
+            ),
+            _ => Err(errors::ConnectorError::FailedToObtainAuthType)?,
```
Comment (SamraatBansal) [inline crates/router/src/connector/stripe.rs:1043]: Correct the error type
```suggestion
            _ => Err(errors::ConnectorError::WebhookReferenceIdNotFound)?,
```

File: crates/router/src/connector/stripe.rs [hunk 9]
```diff
@@ -1026,23 +1027,70 @@ impl api::IncomingWebhook for Stripe {
...
     ) -> CustomResult<api::IncomingWebhookEvent, errors::ConnectorError> {
-        let details: stripe::StripeWebhookObjectEventType = request
+        let details: stripe::StripeWebhookObjectId = request
+            .body
+            .parse_struct("StripeWebhookObjectId")
+            .change_context(errors::ConnectorError::WebhookReferenceIdNotFound)?;
+        let webhook_obj: stripe::StripeWebhookObjectEventType = request
             .body
             .parse_struct("StripeWebhookObjectEventType")
             .change_context(errors::ConnectorError::WebhookEventTypeNotFound)?;
```
Comment (SamraatBansal) [inline crates/router/src/connector/stripe.rs:1058]: You are essentially parsing same struct twice, It can be merged into one and can be used subsequently.
Make a larger struct and parse only once.

File: crates/router/src/connector/stripe.rs [hunk 10]
```diff
@@ -1058,6 +1106,46 @@ impl api::IncomingWebhook for Stripe {
 
         Ok(details.data.object)
     }
+
+    fn get_dispute_details(
+        &self,
+        request: &api::IncomingWebhookRequestDetails<'_>,
+    ) -> CustomResult<api::disputes::DisputePayload, errors::ConnectorError> {
+        let details: stripe::StripeWebhookObjectEventType = request
```
Comment (SamraatBansal) [inline crates/router/src/connector/stripe.rs:1114]: Use a generic name if you are using the same struct multiple times

File: crates/router/src/connector/stripe/transformers.rs [hunk 11]
```diff
@@ -1333,7 +1340,86 @@ pub struct StripeWebhookObjectResource {
...
+    pub id: String,
+    pub amount: i32,
+    pub currency: String,
+    pub reason: String,
+    pub created: i64,
+    pub evidence_details: EvidenceDetails,
+    pub status: DisputeStatus,
```
Comment (SamraatBansal) [inline crates/router/src/connector/stripe/transformers.rs:1360]: the values that are not there always, make those as option, else deserialization will fail

File: crates/router/src/connector/stripe.rs [hunk 12]
```diff
@@ -1051,13 +1092,59 @@ impl api::IncomingWebhook for Stripe {
...
+                .status
+                .ok_or(errors::ConnectorError::WebhookResourceObjectNotFound)?
+                .to_string(),
+            created_at: Some(get_date_and_time(details.event_data.event_object.created)?),
+            updated_at: None,
+        })
+    }
+}
+
+pub fn get_date_and_time(timestamp: i64) -> CustomResult<String, errors::ConnectorError> {
```
Comment (jagan-jaya) [inline crates/router/src/connector/stripe.rs:1136]: create an issue for this

**Review by jagan-jaya — DISMISSED** [review]

File: crates/router/src/connector/stripe.rs (hunk 12 above)
Comment (ShankarSinghC) [reply crates/router/src/connector/stripe.rs:1136]: https://github.com/juspay/hyperswitch/issues/967

File: crates/router/src/connector/stripe.rs [hunk 13]
```diff
@@ -1285,28 +1285,69 @@ impl api::IncomingWebhook for Stripe {
         &self,
         request: &api::IncomingWebhookRequestDetails<'_>,
     ) -> CustomResult<api_models::webhooks::ObjectReferenceId, errors::ConnectorError> {
-        let details: stripe::StripeWebhookObjectId = request
+        let details: stripe::WebhookEvent = request
```
Comment (sai-harsha-vardhan) [inline crates/router/src/connector/stripe.rs:1288]: Can you please rename `WebhookEvent` as it contains both event_type and event_data?

File: crates/router/src/connector/stripe.rs [hunk 14]
```diff
@@ -1285,28 +1285,69 @@ impl api::IncomingWebhook for Stripe {
...
-        let details: stripe::StripeWebhookObjectEventType = request
+        let details: stripe::WebhookEvent = request
             .body
-            .parse_struct("StripeWebhookObjectEventType")
-            .change_context(errors::ConnectorError::WebhookEventTypeNotFound)?;
-
-        Ok(match details.event_type.as_str() {
-            "payment_intent.payment_failed" => api::IncomingWebhookEvent::PaymentIntentFailure,
-            "payment_intent.succeeded" => api::IncomingWebhookEvent::PaymentIntentSuccess,
+            .parse_struct("StripeWebhookObjectId")
```
Comment (sai-harsha-vardhan) [inline crates/router/src/connector/stripe.rs:1322]: `StripeWebhookObjectId` -> `StripeWebhookEventData`

File: crates/router/src/connector/stripe.rs [hunk 15]
```diff
@@ -1285,28 +1285,69 @@ impl api::IncomingWebhook for Stripe {
...
+            .change_context(errors::ConnectorError::WebhookReferenceIdNotFound)?;
+        Ok(match details.event_type {
+            stripe::WebhookEventType::PaymentIntentFailed => {
```
Comment (sai-harsha-vardhan) [inline crates/router/src/connector/stripe.rs:1325]: Redundant {}

File: crates/router/src/connector/stripe.rs [hunk 16]
```diff
@@ -1285,28 +1285,69 @@ impl api::IncomingWebhook for Stripe {
...
+                api::IncomingWebhookEvent::PaymentIntentFailure
+            }
+            stripe::WebhookEventType::PaymentIntentSucceed => {
+                api::IncomingWebhookEvent::PaymentIntentSuccess
+            }
+            stripe::WebhookEventType::DisputeCreated => api::IncomingWebhookEvent::DisputeOpened,
+            stripe::WebhookEventType::DisputeClosed => api::IncomingWebhookEvent::DisputeCancelled,
+            stripe::WebhookEventType::DisputeUpdated => {
+                match details.event_data.event_object.status {
```
Comment (sai-harsha-vardhan) [inline crates/router/src/connector/stripe.rs:1334]: Can we have a `try_from` for this conversion?

File: crates/router/src/connector/stripe.rs [hunk 17]
```diff
@@ -1315,13 +1356,43 @@ impl api::IncomingWebhook for Stripe {
...
         Ok(details.data.object)
     }
+    fn get_dispute_details(
+        &self,
+        request: &api::IncomingWebhookRequestDetails<'_>,
+    ) -> CustomResult<api::disputes::DisputePayload, errors::ConnectorError> {
+        let details: stripe::WebhookEvent = request
+            .body
+            .parse_struct("StripeWebhookObjectEventType")
+            .change_context(errors::ConnectorError::WebhookEventTypeNotFound)?;
```
Comment (sai-harsha-vardhan) [inline crates/router/src/connector/stripe.rs:1373]: Please throw the appropriate errors

File: crates/router/src/connector/stripe.rs [hunk 18]
```diff
@@ -1315,13 +1356,43 @@ impl api::IncomingWebhook for Stripe {
...
+            challenge_required_by: details
+                .event_data
+                .event_object
+                .evidence_details
+                .map(|payload| payload.due_by),
+            connector_status: details
+                .event_data
+                .event_object
+                .status
+                .ok_or(errors::ConnectorError::WebhookResourceObjectNotFound)?
```
Comment (sai-harsha-vardhan) [inline crates/router/src/connector/stripe.rs:1390]: You need not throw error if connector_status is missing

File: crates/router/src/connector/stripe/transformers.rs [hunk 19]
```diff
@@ -1311,6 +1312,8 @@ pub struct RefundRequest {
     pub metadata_txn_id: String,
     #[serde(rename = "metadata[txn_uuid]")]
     pub metadata_txn_uuid: String,
+    #[serde(rename = "metadata[refund_id]")]
+    pub metadata_refund_id: String,
```
Comment (sai-harsha-vardhan) [inline crates/router/src/connector/stripe/transformers.rs:1316]: Are these changes required?

File: crates/router/src/connector/stripe/transformers.rs [hunk 20]
```diff
@@ -1573,34 +1578,119 @@ impl<F, T>
...
-pub struct StripeWebhookDataId {
-    pub object: StripeWebhookDataObjectId,
+pub struct WebhookEventObjectResource {
+    pub data: WebhookEventDataResource,
 }
 
 #[derive(Debug, Deserialize)]
-pub struct StripeWebhookDataResource {
-    pub object: serde_json::Value,
+pub struct WebhookEvent {
```
Comment (sai-harsha-vardhan) [inline crates/router/src/connector/stripe/transformers.rs:1591]: `WebhookEvent` -> `WebhookEventdata`

**Review by sai-harsha-vardhan — CHANGES_REQUESTED** [review]

File: crates/router/src/connector/checkout/transformers.rs [hunk 21]
```diff
@@ -551,3 +552,109 @@ impl From<CheckoutRedirectResponseStatus> for enums::AttemptStatus {
...
+pub struct CheckoutWebhookBody {
+    #[serde(rename = "type")]
+    pub txn_type: CheckoutTxnType,
+    pub data: CheckoutWebhookData,
+    #[serde(default, with = "common_utils::custom_serde::iso8601::option")]
+    pub created_on: Option<PrimitiveDateTime>,
+}
+#[derive(Debug, Deserialize, strum::Display, Clone)]
+#[serde(rename_all = "snake_case")]
+pub enum CheckoutTxnType {
```
Comment (SanchithHegde) [inline crates/router/src/connector/checkout/transformers.rs]: Would be great if we could call this `CheckoutTransactionType` instead. Same for the field in `CheckoutWebhookBody`, please change it to `transaction_type`. We avoid abbreviations like `transaction -> txn` as much as possible in our code, and use verbose names instead.

File: crates/router/src/types/api/disputes.rs [hunk 22]
```diff
@@ -14,7 +15,7 @@ pub struct DisputePayload {
     pub connector_dispute_id: String,
     pub connector_reason: Option<String>,
     pub connector_reason_code: Option<String>,
-    pub challenge_required_by: Option<String>,
-    pub created_at: Option<String>,
-    pub updated_at: Option<String>,
+    pub challenge_required_by: Option<PrimitiveDateTime>,
+    pub created_at: Option<PrimitiveDateTime>,
+    pub updated_at: Option<PrimitiveDateTime>,
```
Comment (SanchithHegde) [inline crates/router/src/types/api/disputes.rs:22]: If this type is implementing `Deserialize`, it's best to provide an expected format, such as ISO 8601.

File: crates/storage_models/src/dispute.rs [hunk 23]
```diff
@@ -17,14 +17,14 @@ pub struct DisputeNew {
...
     pub connector_status: String,
     pub connector_dispute_id: String,
     pub connector_reason: Option<String>,
     pub connector_reason_code: Option<String>,
-    pub challenge_required_by: Option<String>,
-    pub dispute_created_at: Option<String>,
-    pub updated_at: Option<String>,
+    pub challenge_required_by: Option<PrimitiveDateTime>,
+    pub dispute_created_at: Option<PrimitiveDateTime>,
+    pub updated_at: Option<PrimitiveDateTime>,
```
Comment (SanchithHegde) [inline crates/storage_models/src/dispute.rs:26]: Same here.

File: crates/storage_models/src/dispute.rs [hunk 24]
```diff
@@ -44,9 +44,9 @@ pub struct Dispute {
...
```
Comment (SanchithHegde) [inline crates/storage_models/src/dispute.rs:49]: And here.

File: crates/router/src/connector/stripe/transformers.rs [hunk 25]
```diff
@@ -1573,34 +1588,119 @@ impl<F, T>
...
+    #[serde(rename = "data")]
+    pub event_data: WebhookEventData,
 }
 
 #[derive(Debug, Deserialize)]
-pub struct StripeWebhookObjectResource {
-    pub data: StripeWebhookDataResource,
+pub struct WebhookEventData {
+    #[serde(rename = "object")]
+    pub event_object: WebhookEventObjectData,
```
Comment (SanchithHegde) [inline crates/router/src/connector/stripe/transformers.rs:1611]: The `#[serde(rename = "...")]` annotation is typically used when the field has the same name as a Rust keyword (for example `type`, `trait`, etc.). In other cases, it's not needed, but can be added to improve readability.

Here, it _can_ be removed for all fields excluding `event_type`, since they don't alter the readability by a great extent. It's not a necessary change, I'll leave it upto you to take a call on this.

**Review by SanchithHegde — COMMENTED** [review]
Other than that, looks good to me!

**Review by SanchithHegde — APPROVED** [review]

**Review by sai-harsha-vardhan — APPROVED** [review]
LGTM

**Review by SanchithHegde — APPROVED** [review]

**Review by SamraatBansal — APPROVED** [review]

**Review by jarnura — APPROVED** [review]

**Review by jarnura — CHANGES_REQUESTED** [review]
Add required labels.

**Review by jarnura — APPROVED** [review]
